### PR TITLE
Property expression highlight

### LINKED ISSUE #40: Highlight support for property expression
When cursor select a property expression value or a property:

```
#  A test.property
test.property=hello
other.property=${test.pro|perty}, world
```

It should highlight the test.property property name and the ${test.property}

### PR DESCRIPTION
**Pull request by datho7561**
Highlights references when cursor is on a property key,
and the property key when the cursor
is on a reference (property expression).

Closes #40

Requires [the property expressions PR to be merged first](https://github.com/eclipse/lsp4mp/pull/13)

- [x] Tests

Signed-off-by: David Thompson <davthomp@redhat.com>

### CODE REVIEW

**Comment by datho7561** [comment]
Demo:

DocumentHighlightDemo

I modified `editor.selectionHighlightBackground` under `workbench.colorCustomizations` to make it more visible.

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/MicroProfileDocumentHighlight.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+public class MicroProfileDocumentHighlight {
+
+	/**
+	 * Returns a list of highlights for a MicroProfile properties document
+	 *
+	 * @param document The MicroProfile properties document
+	 * @param position The position of the cursor
+	 * @return a list of highlights for a MicroProfile properties document
+	 */
+	public List<? extends DocumentHighlight> doDocumentHighlight(PropertiesModel document, Position position) {
```
Comment (angelozerr) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/MicroProfileDocumentHighlight.java:41]: I will rename to `findDocumentHighlights` to be consistent with XML Language Service, and TypeScript vscode-html language service, json, etc

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/MicroProfileDocumentHighlight.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+			PropertyValueExpression node) {
+		String otherProp = node.getReferencedPropertyName();
+		if (otherProp == null || otherProp.isBlank()) {
+			return Collections.emptyList();
+		}
+		for (Node child : document.getChildren()) {
+			if (child.getNodeType() == NodeType.PROPERTY) {
+				Property property = (Property) child;
+				if (otherProp.equals(property.getPropertyName())) {
+					return Collections
```
Comment (angelozerr) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/MicroProfileDocumentHighlight.java:76]: You should highlight the property expression which have the cursor as `DocumentHighlightKind.Read` too.

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/MicroProfileDocumentHighlight.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+		}
+		List<DocumentHighlight> highlights = new ArrayList<>();
+		for (Node child : document.getChildren()) {
+			if (child.getNodeType() == NodeType.PROPERTY) {
+				Property property = (Property) child;
+				if (property.getValue() != null) {
+					for (Node valueSegment : ((Property) child).getValue().getChildren()) {
+						if (valueSegment.getNodeType() == NodeType.PROPERTY_VALUE_EXPRESSION && propertyName
+								.equals(((PropertyValueExpression) valueSegment).getReferencedPropertyName())) {
+							highlights.add(new DocumentHighlight(PositionUtils.createRange(valueSegment), DocumentHighlightKind.Read));
```
Comment (angelozerr) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/MicroProfileDocumentHighlight.java:107]: You should highlight the referenced property which have the cursor as `DocumentHighlightKind.Write` too.

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/MicroProfileDocumentHighlight.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	/**
+	 * Highlight the key portion of the assignment statement for the property that
+	 * the property expression references, or nothing
+	 *
+	 * @param document The properties model that is being worked with
+	 * @param node     The property value expression that the cursor is on
+	 * @return A singleton list that contains a highlight for the key in the
+	 *         assignment statement for the reference property, or an empty list
+	 */
+	private List<? extends DocumentHighlight> propertyValueExpressionHighlight(PropertiesModel document,
```
Comment (angelozerr) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/MicroProfileDocumentHighlight.java:66]: In Java land, we use `get` prefix for method which return something, so please rename to `getPropertyValueExpressionHighlight`

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/MicroProfileDocumentHighlight.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	 * Highlight all the property expressions that reference the property that the
+	 * given property key represents
+	 *
+	 * @param document The properties model that is being worked with
+	 * @param node     The property key that the cursor is on
+	 * @return A list containing highlights for all property expressions in the
+	 *         document that reference the expression that the property key
+	 *         represents
+	 */
+	private List<? extends DocumentHighlight> propertyKeyHighlight(PropertiesModel document, PropertyKey node) {
```
Comment (angelozerr) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/MicroProfileDocumentHighlight.java:94]: please rename to `getPropertyKeyHighlight`

**Comment by fbricon** [comment]
@datho7561 please make sure new file headers follow this format:

> /********************************************************************************
>  * Copyright (c) {date} {owner}[ and others]
>  * 
>  * This program and the accompanying materials are made available under the
>  * terms of the Eclipse Public License v. 2.0 which is available at
>  * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
>  * which is available at https://www.apache.org/licenses/LICENSE-2.0.
>  * 
>  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
> 
> ********************************************************************************/

**Comment by angelozerr** [comment]
@datho7561 please fix conflicts

**Comment by datho7561** [comment]
rebased. The property expression PR will need to be merged before this, and this will need to be re-rebased

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/MicroProfileDocumentHighlight.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	/**
+	 * Returns a DocumentHighlight that highlights the given node
+	 *
+	 * @param n the node to highlight
+	 * @param kind the kind of the DocumentHighlight
+	 * @return a DocumentHighlight that highlights the given node
+	 */
+	private static DocumentHighlight getDocHighlight(Node n, DocumentHighlightKind kind) {
```
Comment (angelozerr) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/MicroProfileDocumentHighlight.java:130]: I will rename to `createHighlight`

**Pull request merged**